### PR TITLE
Bugfix settimeout warning

### PR DESCRIPTION
**Pull request by Meiguro**
@sarfata quite a bit of this is improving stack trace translation. The notable changes however are the new one-time warnings of setTimeout and setInterval. Quite a few people have been getting hit by this JavaScript gotcha. Do you have any comments for the warning messages?

### CODE REVIEW

File: src/js/lib/safe.js [hunk 1]
```diff
@@ -159,26 +166,47 @@ Pebble.sendAppMessage = function(message, success, failure) {
...
 navigator.geolocation.getCurrentPosition = function(success, error, options) {
   return getCurrentPosition.call(this, safe.protect(success), safe.protect(error), options);
 };
 
+var ajax;
+
+/* Try to load the ajax library if available and silently fail if it is not found. */
+try {
+  ajax = require('ajax');
+} catch (err) {}
```
Comment (sarfata) [inline src/js/lib/safe.js]: ?

When does that happen? Why do we need the try/catch?

File: src/js/lib/safe.js [hunk 2]
```diff
@@ -159,26 +166,47 @@ Pebble.sendAppMessage = function(message, success, failure) {
 /* Wrap setTimeout and setInterval */
 var originalSetTimeout = setTimeout;
 window.setTimeout = function(callback, delay) {
+  if (safe.warnSetTimeoutNotFunction !== false && typeof callback !== 'function') {
```
Comment (sarfata) [inline src/js/lib/safe.js]: What is the default JS behavior? Why do we need to do this?

Error message looks fine to me. I am just surprised that we have to do this.

File: src/js/lib/safe.js [hunk 3]
```diff
@@ -44,15 +36,21 @@ var makeTranslateStack = function(stackLineRegExp, translateLine) {
         line = lines[i] = translateLine.apply(this, m);
       }
       if (line.match(module.filename)) {
-        lines.splice(--i, 2);
+        lines.splice(i, 1);
```
Comment (sarfata) [inline src/js/lib/safe.js]: hard to review those changes without knowing what the problems were. In general I would say this is the kind of functions that would deserve unit test. Or at least clear comments on expected in/out.

File: src/js/lib/safe.js (hunk 1 above)
Comment (Meiguro) [inline src/js/lib/safe.js]: pebble.js is a monolithic framework at the moment, but many of its parts can be used in other projects. I want to start modularizing every component. At some point, I would like to be able to recommend using safe.js + loader.js + wscript in someone's C + JS project.

Comment (Katharine) [inline src/js/lib/safe.js]: That's good, because that's coincidentally my plan.

File: src/js/lib/safe.js (hunk 2 above)
Comment (Meiguro) [inline src/js/lib/safe.js]: The default behavior for JavaScript is to silently fail and return a new timer ID. This confuses many developers using pebble.js, so I decided to put a one time warning that can be disabled if really necessary.

File: src/js/lib/safe.js (hunk 3 above)
Comment (Meiguro) [inline src/js/lib/safe.js]: Agreed, there should be unit tests. This used to remove two lines that were useless, but that's tricky and not always the case. Now it removes the current line if it is safe.js, which is much more readable.

**Comment by sarfata** [comment]
:+1:

**Pull request merged**